### ISSUE TITLE
Issues when using Ajax.ActionLink to load partial views

### ISSUE THREAD

**Issue by kkfrosty** [IdentityServer/IdentityServer3#2466]
We have an existing application which we're just starting to convert over to where our menu items utilize Ajax.ActionLink to load all views in our application as partial views.

Everything was working or so we though.  But after approx. 3 minutes or more and it's not exact every time, we've been noticing weird behavior.

First, we were getting cross orgin errors that Access-Control-Allow-Origin was not present.  Even though the domain for our app and identity are the same.  For testing it's localhost.

In troubleshooting we noticed that there was a http call to our identity server to load a script.  The server site responding that it would support calls to HTTP.  So we converted our app to load to https which matches Identity requirements.

Now, when we load the app, we can load our converted views with ajax.  But again if we pause for a couple of minutes, now the views load missing the _Layout template.  Thus all css, scripts etc are gone.

In using F12, I notice the same pattern each time.  An XMLHttpRequest call is made to the view but a 401 is returned.  It makes the exact same call again and this time a 302 is returned.

Then we see a call to authorize mad to our identity server which comes back ok.
Then a call is made to app.FormPostResponse.JS on identity server which comes back ok.

Then 3 calls are made to the root URL of our application in the form of post's.  The first 2 come back with 401 responses.  The next comes back with a 302.

Then a XMLHttpRequest is made again to the partial view and comes back OK.  But for some reason a 2nd call is made and that gets hung in a Pending state.

At this point the partial view data shows on the page, unformatted in any way because scripts and css from Layout are completely missing.

Right now we're stumped.  We just know that after these couple of minutes of inactivity, any call Identity tries to authorize.

Looks like it does but the last XMLHttpRequest gets hung in a pending state.

Here is a copy F12 tools screen shot that shows the sequence of events.

![ajaxidentity](https://cloud.githubusercontent.com/assets/10908749/12517959/91637878-c103-11e5-92f6-ed274aa989aa.png)

Any ideas of where or what to look at next would be appreciated.

**Comment by brockallen** [comment]
Related? http://brockallen.com/2013/10/27/using-cookie-authentication-middleware-with-web-api-and-401-response-codes/

**Comment by kkfrosty** [comment]
I thought it was initially was working but I waited another 5 minutes before and clicked one of our links and again pretty much the same thing. 

I'm running everything from IIS on Windows 10.  I set breakpoints in the OnApplyRequest your link talks about and I never see if get hit.  

Also I set in UseOpenIdeConnectAuthentication in the AuthorizationCodeReceived.

This break point is getting hit on this sequence of events so their is authentication going on again and it looks to complete successfully.

But again you see a call to XMLHttpRequest and you get back a 200, but then it makes the call immediately again and never goes out of a pending state.

For what it's worth, in the browser address, you see the entire ?X-Request-With=XMLHttpRequest&_145......  If you just copy and paste that into a browser window then all you get is the partial back without css or scripts loaded from layout.

Also, data is coming back from the partial.  The controller actions are protected by claim attributes so the caller must have permissions or that should fail.

**Comment by kkfrosty** [comment]
For what it's worth, we're also having to use the Kentor Owin Cookie Saver because we had nonce http 400 errors randomly a while back until we implemented it in ConfigureAuth

**Comment by kkfrosty** [comment]
Testing this more, the simple version it appears that when we click on an Ajax.ActionLink and there is a  Redirect made for AuthenticateRequest, then some how in the process, we're losing _Layout which contains all our script and css references.

The redirect is working because the data from our partial page is making it to the screen, but with all the script and css references missing, we just see text on the screen.

If we click back so the home page reloads again which isn't loaded via ajax, then all works again and there isn't another authentication request either.

**Comment by kkfrosty** [comment]
Update: Sunday morning I tried testing some more.  Nothing changed that I know of but now pages are loading on my dev machine.  Windows 10 via IIS.  (Well I added the OnApplyRedirect but the break point is never getting hit so I don't think that has any bearing.

However, 3 other developer machines they are still having the problem.  Again, we load our app and if you leave the app idle for approx. 3 minutes or more, then you click on a link.  It appears to be triggering an authenticationrequest.

When that happens you see two calls ajax calls to load a partial view.  One comes back the other is pending.

The end result is we lose _Layout from the partial view.  (Note you can't define Layout in the partial as it causes all kinds of other problems.  Even if we could the whole idea of using ajax is to prevent everything from reloading.)

If no ideas, then a couple of questions I have.

The ticket we get from Identity shouldn't be expired so why the authentication request after only a couple of minutes.

Even on my machine that works, I notice after a few minutes, calls to any thing first result in a 401 and then you see the call again that succeeds.  This is even after re authentication which was successful.  So basically we're seeing duplicate traffic.

ConfigAuth we're using UseCookieAuthentication as well as useOpenIdConnectAuthentication which points to our IdentityServer.

In addition to any insight, does anybody have any ideas as to how to go about troubleshooting MVC to figure out why we're seeing these authentication requests to start with and what is causing them after only a few minutes?

**Comment by brockallen** [comment]
In the MVC app, on the OIDC middleware do you have `UseIdentityTokenLifetime=false`? If you don't, then your cookie will only be good for 5 minutes in the MVC app. This quick expiration might be explained if you're missing that setting.

**Comment by kkfrosty** [comment]
Thank you.  We do not have that set.  In our UseOpenIdConnectAuthentication, I only see UseTokenLifeTime.

I set this to false.  Reset IIS and cleared cache and I don't see any change to what takes place in F12 Network for IE

**Comment by brockallen** [comment]
Yea, that's the property I meant -- `UseTokenLifeTime=false`, My prior message had a typo.

**Comment by brockallen** [comment]
In essence, you're making an ajax call that authenticates based on cookies. If the cookie is no longer valid (or the identity in the cookie is not allowed) then the app is triggering redirect to IdSvr for authentication. Does this sound like what's happening?

**Comment by kkfrosty** [comment]
We are using Cookies in our authentication.
In theory what you state sounds valid.  But why would the cookie become invalid in several minutes?  (We don't know the exact time we have to leave the pages idle, we've waited as little as 3 minutes but it could have been 5.)

If we are clicking on things in the application then the authentication request doesn't happen.

Right now we only have 2 views loading via ajax and 75+ others loading via type MVC non ajax methods and haven't noticed the problem.

Note, 4 of our developers still are having the original problem.  Meaning when the Authentication happens again, they lose _Layout and the last call just stays in a pending state.  I have to believe there isn't an authentication issue as for they get the data back from the db's via the controller which require proper authentication.

For me that particular problem went away but I have no idea why.  I'm bringing up another VM to test.

At this time all I have to go on is what I see in F12 network and trying to figure out what else we can do to troubleshoot the authentication.  We understand it has to happen at some point but trying to figure out where that's controlled and if we can increase it on my machines where there are no problems.

Hopefully in going down this route we'll also figure out the problem on our developer workstations where _Layout goes missing.

**Comment by brockallen** [comment]
> But why would the cookie become invalid in several minutes?

Not sure, so check the cookie expiration.

> In theory what you state sounds valid.

The reason I wanted to be certain is that the blog post I liked before will help you solve this. The way cookie authentication works assumes you have a full browser window that's making the request, and so when the cookie expires you can redirect the user to go login. As you found, that doesn't work with Ajax calls. So what you need to do is return 401 on ajax calls when the authorization fails (so no 302). 

As for why the cookie is expiring so quickly, I don't know.

**Comment by kkfrosty** [comment]
I found OnValidateIdentity and when logging in, the cookie's expiration looks to be 2 weeks.

It appears we can change this by setting ExpireTimeSpan in UseCookieAuthentication.

If the cookie hasn't expired, then the other thing that would cause authentication would be an invalid cookie correct?

If that's the case, is there anyway to subscribe to an event to where I can tell what the exact problem is an try to fix it.

You are correct I'm not seeing any 302's anymore.  What's odd is the OnApplyRedirect in the blog post you mentioned, I don't think the code is ever running.

I had set breakpoints but they never hit.  So now I throw an exception in it and that never happens either.

I'm not seeing a 302 anymore on my machine but the OnApply code from the blog isn't running.  I actually took the code out for testing and not seeing the 302's.

I'm kind of at a loss as to why.

So I'm left with trying to figure out how we can troubleshoot to see what's forcing the Authentication when we know the cookie expire date hasn't passed.

Also, on a hand full of developer machines that still are having the original issue. First why the authentication is happening and then whatever is causing the us to loose _Layout.

**Comment by kkfrosty** [comment]
Update, for some reason our developer workstations now appear to not be seeing the 302's and that part doesn't appear to be a problem anymore.

Again we don't know why.  One thing, when they were getting the 302's they were seeing a call to get a js script off our identity server.  Even though no where in our code or configs is there a URL to identity that used HTTP, the call was being made with HTTP.  That was throwing an exception pertaining to HTTPS must be used on the server.

In dealing with that, we changed the redirect URL back to our application to use HTTPS.  When I did that I started getting the 302's.

I added OnApply as Brock suggested but I've never seen that break point hit and have actually removed the code in testing.  

Regardless of what stopped the 302's, now all we're trying to understand and how to troubleshoot is why our authentication cookie after several minutes is not being accepted and we're seeing a 401 causing authentication.

I've added these to UseCookieAuthentication

                CookieHttpOnly = false,
                CookieName = "OurAuthenticationCookie",
                CookieSecure = CookieSecureOption.Always,
                SlidingExpiration = true,

But I haven't seen them make a difference.  We're thinking it has to be something with the cookie because our understanding either a bad cookie or an expired cookie will cause the authentication.  From what we see, the cookie isn't expired.

**Comment by brockallen** [comment]
> If the cookie hasn't expired, then the other thing that would cause authentication would be an invalid cookie correct?

Or the user is not authorized.

**Comment by kkfrosty** [comment]
How would they just become unauthorized in a couple of minutes.  I checked and the ticket expiration is an hour right now.

Either way, is there a good way to subscribe to an event, set a break point somewhere to figure out what the case may be?

**Comment by brockallen** [comment]
For the Ajax call to load UI, not sure. But if any ajax calls decides the user is not allowed to do the operation they're requesting, it should return 401 (actually 403).

**Comment by kkfrosty** [comment]
Our controllers we have a custom claims access attribute we use.  When the app loads we load AuthorizeAttribute.  I created a custom one and if I get the user, IsAuthenticated is true.  The user has the claims needed to authorize for the controller call.  I also over rode HandleUnauthorizedResult here as well and the breakpoint there isn't getting hit.

Our custom ClaimsAccessAttribute has an override for HandleUnauthorizedRequest and I'm not seeing that fire.

Those are the places we're forcing authorization and I don't see the HandleUnauthorizedRequest sections running.

One thing I have just learned, when whatever happens, the 401 is occurring somewhere before the AuthorizationAttribute or ClaimsAccessAttribute code is utilized.  Because in F12 network, I see a 401 and when my breakpoints start getting hit in our authorize and claim attributes, it's the second call and the 401 has already occurred.

**Comment by brockallen** [comment]
Any update?

**Comment by kkfrosty** [comment]
No unfortunately not.  I enabled IIS logging and I didn't see the 401 that I see in F12.  

As mentioned we set break points HandleUnAuthorizedRequest overrides for our custom AuthorizeAttribute and custom ClaimAttribute and those break points never get hit.

At this point we're not having the _Layout issue that caused us to stop work.  Now the 401's after a couple of minutes are just something we notice from troubleshooting the other problem. 

We had to start back to work.  I was hoping somebody may have some ideas along the way here we can eventually use to get a handle on the 401.  Otherwise we're moving on until we have spare time to experiment.

**Comment by brockallen** [comment]
So again, I'm really confused why you're expecting a 302 from your MVC app to IdentityServer to work from an Ajax call. Do you understand what I'm getting at, or am I missing something?

**Comment by kkfrosty** [comment]
We don't expect the 302 to work.  Those have went away and now all we're trying to do is figure out why we see 401's after a few minutes of inactivity.  Everything is working but wasted calls and overhead.  In development not an issue but if we can figure out the calls before production it would be nice.

We've established via you and this thread that this would happen if the cookie is corrupt, timed out or user wasn't authenticated correct?

Again, to troubleshoot the 401's we've made sure the user is authenticated by the things I've listed.  We've set break points in the HandleUnauthorizedRequest overrides I've mentioned.

All we're trying to do now is figure out why we see the 401's.  Pages are working now, just trying to cut out any unnecessary overhead if we can.

**Comment by brockallen** [comment]
Ah, ok... so now it's just that you're failing to authenticate after a short amount of time. Ok.

So maybe a load balancing issue if it's only in production? One server doesn't accept the other server's cookie? Did you sync the `<machineKey>`?

**Comment by kkfrosty** [comment]
We just see the 401 packets after a couple of minutes of inactivity.  But everything seems to be working.  Looking at the Identity Logs, you see an authenticate based on Subject if I recall correctly. So right now nothing is ultimately failing like it was with the 302's.

It's just after 3+ minutes or whatever the time is, we see the 401, you have the authentication process but then you start seeing duplicate calls for example calls for css, script files etc because even after the authentication, the first call to such things gets a 401 returned.

We do not have this in production.  What I'm trying to say, if we can cut out these duplicate calls, we'd like to BEFORE this goes into production.

Right now we're all running this on our dev workstations via IIS.  

At this point if timing weren't an issue I would say everything is working as it should.  But we'd like to control when the 401 takes place if it were a cookie timing out which doesn't seem to be the case.

We thought/think it's something to do with the cookie.  But maybe it's a IIS connection timeout or such.

It's why I asked if there an event we can subscribe to somewhere or logging we could enable etc to find out the real reason the 401 happening after minutes when the user is authenticated from everything we've tested.

**Comment by brockallen** [comment]
I can't say since that would mainly be debugging your MVC app, and it's hard for me to know what's going on there.

Let us know when you find out more.

**Comment by kkfrosty** [comment]
Understand, we really appreciate your help.  The show stopper issue was resolved.

As for the 401, our app I think is pretty simple as far as authentication goes.  

We use app.UseOpenIdConnectAuthentication pretty much the same as we found in your of the Identity samples.

We have a custom Authorization Attribute and a custom Claims Attribute.  We've set break points in those and don't see any issues or calls to things like HandleAuthorizeRequest as we would think if there was an authorization issue. 

Whatever is going on we think is in MVC related source and if anybody has any pointers on a custom attribute we could utilize and override any methods or such to help determine why after a couple of minutes we see a 401 and if there is a setting to control the frequency, we'd greatly appreciate it.
